### PR TITLE
Expandable commit bodies

### PR DESCRIPTION
**Pull request by sondr3**
## UPDATE: Here is how it looks right now:

e5998be9e0d7352e72b44f2918ce2f2a
f48f9d6f52849b35644d0b53791b964d

## Original post
I've taken some time to try to work on some kind of solution to the expandable commit message bodies that was talked about in #2939 and #2944, but trying to make it work like it does on GitHub and GitLab. It is however a very rough work in progress as I've never programmed in Go before nor have I much experience with JQuery.

There are several issues with this right now:

- ~It shows the expand button for every commit, even if it's not required.~
- ~The commit bodies doesn't expand properly, the text get squished together in one line and I'm not sure how to display it over several lines~
- ~There's quite a bit of code duplication going on with the commit bodies~

I'd like a helping hand with this to figure out how to best solve it, as I'm not a Go programmer.

### CODE REVIEW

**Comment by lafriks** [comment]
You should add function that checks if message contains newlines and returns true if it does and false otherwise. This way you can add:
```
{{if IsMultilineCommitMessage .Message }}
<button ....
<p class="commit-body" ....
{{end}}
```

**Comment by sondr3** [comment]
Where can I find how `.Message` works? It seems it always contains a newline and it makes my function always return true.

**Comment by sondr3** [comment]
Okay, I've fixed all the commit having the box for showing the commit body, now I've just gotta figure out how to properly render it in HTML.

File: modules/templates/helper.go
Comment (lafriks) [inline modules/templates/helper.go]: just do
```
return strings.Count(strings.TrimSpace(msg), "\n") > 1
```

**Review by lafriks — CHANGES_REQUESTED** [review]

File: modules/templates/helper.go
Comment (sondr3) [reply modules/templates/helper.go]: Sweet, I was looking for a better way to do that.

**Comment by lafriks** [comment]
Add comments to public functions

**Comment by sondr3** [comment]
I've managed to make it place the commit bodies somewhat properly into the HTML now, my biggest hurdle now is getting the cells to stop moving when you expand the commit bodies. The image below shows what happens, when you press the [...] button it moves the commit time, avatar, username and SHA-code into the middle of the cell, which makes the whole layout jump around when you press it.

How would I stop that from happening?

screen shot 2017-11-26 at 23 41 27

Also, where do we display the commit messages in Gitea? I need to update all the places where we show the commit message to this new format.

@lafriks yeah, I haven't added them yet. I'll do it once I'm happy with them.

**Comment by lafriks** [comment]
To stop jumping you need to add `vertical-align: top` to css

File: modules/templates/helper.go
Comment (ethantkoenig) [inline modules/templates/helper.go]: I think we should differentiate between single and double new-lines, since a single newline may just be wrapping a long line, but a double newline typically indicates the start of a new paragraph. In fact, I think the best solution is just to render newlines as they appear in the commit message (this is what Github does).

~Also, note that newlines in a `template.HTML` will *not* be rendered as newlines. The somewhat hacky way to fix this problem is to replace newlines with `<br>`.~ (the crossed-out part is incorrect, ignore it)

File: public/js/index.js
Comment (ethantkoenig) [inline public/js/index.js]: I'm not a JS expert, but this seems fragile. What if we add a DOM element between the commit button and the commit body? Perhaps it would be better to select for the `commit-body` class?

File: templates/repo/view_list.tmpl
Comment (ethantkoenig) [inline templates/repo/view_list.tmpl]: Please indent with tabs

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: templates/repo/view_list.tmpl
Comment (ethantkoenig) [inline templates/repo/view_list.tmpl]: This should not be inside a paragraph, but instead inside a `<pre>` like in the other template (in order for newlines to be rendered properly).

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: templates/repo/view_list.tmpl
Comment (sondr3) [reply templates/repo/view_list.tmpl]: Yep, I forgot to change this one once I updated the other.

File: templates/repo/view_list.tmpl
Comment (sondr3) [reply templates/repo/view_list.tmpl]: I'll fix it, Golands for some reason doesn't want to respect the settings.

File: public/js/index.js
Comment (sondr3) [reply public/js/index.js]: I tried doing

```js
$(".commit-button").click(function() {
    $("commit-body").toggle();
```

But that opens and closes all of the commit messages, not just the one selected. All the documentation I could find from both JQuery themselves and online used this method, otherwise I think we'd have to use one of those JQuery-Toggle libraries.

File: modules/templates/helper.go
Comment (sondr3) [reply modules/templates/helper.go]: Yeah, I initially tried joining on `<br>` but then I looked at Github and saw that they wrapped their commit messages in `<pre>`. However, if that's the case I think the method needs to work differently because it already splits on `\n` and then I discard the first element and join the others. It seems to render correctly, but I'll see what I can figure out.

File: public/js/index.js
Comment (lafriks) [reply public/js/index.js]: Try:
```js
$(".commit-button").on('click', function() {
  $(this).parent().find('.commit-body').toggle();
});
```

Comment (sondr3) [reply public/js/index.js]: Thanks, I'll give it a go later today.

File: modules/templates/helper.go
Comment (lafriks) [inline modules/templates/helper.go]: This function does not seems to be used anywhere, so should be dropped

**Review by lafriks — CHANGES_REQUESTED** [review]

File: modules/templates/helper.go
Comment (sondr3) [reply modules/templates/helper.go]: Sounds good! I just copied the `RenderCommit` functions and changed them a tiny bit.

**Comment by sondr3** [comment]
I've updated the code with the suggestions from you guys, but I also changed the `vertical-align` setting from `top` to `baseline`, otherwise the text gets ever so slightly bumped upwards.

edit: I'm not sure what the error is now, it suggests adding a newline to the CSS as far as I can see but that didn't work.

**Comment by sondr3** [comment]
I've fixed a few other issues, however there are some things I want to change but I'm not sure whether they are out of the scope of this request, namely properly listing commits in issues and pull requests, nor whether I'll have time.

**Comment by lafriks** [comment]
Do only one thing in PR. It's better to submit multiple PRs. You also need to do `make generate-stylesheet` to update index.css

**Comment by sondr3** [comment]
Like this? I ran `make generate-stylesheet` and ran `make stylesheet-check` locally and it passes there.

**Comment by lafriks** [comment]
@sondr3 yes. Now you need to add missing comments:
```
/srv/app/src/code.gitea.io/gitea/modules/templates/helper.go:283:1: comment on exported function RenderCommitBody should be of the form "RenderCommitBody ..."
/srv/app/src/code.gitea.io/gitea/modules/templates/helper.go:301:1: exported function IsMultilineCommitMessage should have comment or be unexported
```

**Comment by sondr3** [comment]
Weird, I must've looked at an old build log by accident because it gave an error on the stylesheets. I'll fix it tomorrow.

**Comment by sondr3** [comment]
I've added comments like the output said but I still got the same message and locally I get a bunch of totally different errors. Any ideas?

**Comment by sondr3** [comment]
Alright, the tests are passing. I had to force push the latest commit again for it to properly test it for some reason, but hey. If anybody wants to give this a go I'd appreciate it! Should I squash the commits together?

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
LGTM

**Comment by lunny** [comment]
LGTM

**Comment by lunny** [comment]
@ethantkoenig needs your approval.

File: public/less/_repository.less
Comment (ethantkoenig) [inline public/less/_repository.less]: Please indent with tabs

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: templates/repo/commits_table.tmpl
Comment (techknowlogick) [inline templates/repo/commits_table.tmpl:66]: There is inline CSS here, could you change it out with a CSS class?
Also this requires JS which means if someone has JS disabled then they won't be able to see the long commit message (perhaps default to open when JS is disabled?)

File: templates/repo/view_list.tmpl
Comment (techknowlogick) [inline templates/repo/view_list.tmpl:33]: same as above regarding inline CSS

File: templates/repo/commits_table.tmpl
Comment (techknowlogick) [reply templates/repo/commits_table.tmpl:66]: Or perhaps hide the button if the user doesn't have JS is another option

Comment (lafriks) [reply templates/repo/commits_table.tmpl:66]: @techknowlogick I think for this inline style is ok

Comment (sondr3) [reply templates/repo/commits_table.tmpl:66]: I agree it's not the prettiest, but the `toggle` function changes the style inline so I thought it was a decent compromise. Regarding what to do when JS is disabled, I'll leave that to the Gitea maintainers, I think Gitea depends on JS enough that if you disable it you're on your own.

File: public/less/_repository.less
Comment (sondr3) [reply public/less/_repository.less]: Will fix.

**Comment by sondr3** [comment]
Alright, fixed the spaces not being tabs. Given that this pull request contains some 20 commits, should I squash them together or just keep them as is?

**Comment by lunny** [comment]
@sondr3 just keep them as is. We will merge with squash.

**Review by ethantkoenig — APPROVED** [review]

**Comment by ethantkoenig** [comment]
LGTM

**Pull request merged**
